### PR TITLE
Bug Fixes + Support for External Load Balancer

### PR DESCRIPTION
**Pull request by shailesh91**
1. Get FQDN resolves Ip into standard PTR format, it should have resolved either to machine name or just returned an IP. (https://en.wikipedia.org/wiki/Reverse_DNS_lookup#IPv4_reverse_resolution)
2. Portal Unregister resource encoding fixed to UTF8
3. Server Upgrade Configuration, Windows Service Resource wasn't imported and one of the conditions failed due to type mismatch
4. Deprecating Global SSL Certs. Making Node Level Certificates the Standard. 
5. Multiple SSL certs per machine with a target component (can specify multiple targets for the same certificate)
6. Dividing Configure DSC Configuration file into Configure components, Federation and Registering SQL Server.
7. Fixing the endpoint URLs output at end of the job
8. Adding support for External Load Balancers

### CODE REVIEW

File: Modules/ArcGIS/ArcGIS.psm1 [hunk 1]
```diff
@@ -259,53 +247,37 @@ function Get-ArcGISURL
...
-            $PrimaryPortalMachineName = $PrimaryPortalMachine.NodeName
+            $PrimaryPortalMachineNode= $AllNodes[$i]
         }
     }
 
-    if($PrimaryPortalMachine)
+    if($PrimaryPortalMachineNode)
     {
-        $PortalExternalDNSName = Get-FQDN $PrimaryPortalMachineName
+        $PortalExternalDNSName = Get-FQDN $PrimaryServerMachineNode.NodeName
```
Comment (spitzerr) [inline Modules/ArcGIS/ArcGIS.psm1]: Shouldn't this be
` $PortalExternalDNSName = Get-FQDN $PrimaryPortalMachineNode.NodeName `
?

File: Modules/ArcGIS/ArcGIS.psm1 [hunk 2]
```diff
@@ -388,12 +338,12 @@ function Get-ArcGISURL
         $RestURL = "https://$ServerExternalDNSName/$ServerContext/rest"
     }
     
-    if($PrimaryPortalMachine)
+    if($PrimaryServerMachineNode)
```
Comment (spitzerr) [inline Modules/ArcGIS/ArcGIS.psm1]: Shouldn't this be
` if($PrimaryPortalMachineNode) `

?

File: Modules/ArcGIS/DSCResources/ArcGIS_Portal/ArcGIS_Portal.psm1 [hunk 3]
```diff
@@ -548,10 +548,9 @@ function Set-TargetResource {
...
                 $sysProps = @{ }
             }
-            $ExpectedWebContextUrl = "https://$($ExternalDNSName)/$($PortalContext)"
-            $ExpectedPrivatePortalUrl = "https://$($PortalEndPoint):7443/arcgis" # "https://$($ExternalDNSName)/$($PortalContext)" 
-            if (($sysProps.WebContextURL -ine $ExpectedWebContextUrl) -or ($sysProps.privatePortalURL -ine $ExpectedPrivatePortalUrl)) {
-				
+            
+            if ($ExternalDNSName){
+                $ExpectedWebContextUrl = "https://$($ExternalDNSName)/$($PortalContext)"
                 Write-Verbose "One of the system properties for WebContextURL '$($sysProps.WebContextURL)' or privatePortalURL '$($sysProps.privatePortalURL)' does not match expected values"
```
Comment (spitzerr) [inline Modules/ArcGIS/DSCResources/ArcGIS_Portal/ArcGIS_Portal.psm1]: Its not intuitive to me, why 
` if ($ExternalDNSName){ `
results in
`  Write-Verbose "One of the system properties for WebContextURL '$($sysProps.WebContextURL)' or privatePortalURL '$($sysProps.privatePortalURL)' does not match expected values"  `

Probably Debug message Needs to be adjusted.

File: Modules/ArcGIS/Configuration/ArcGISConfigure.ps1 [hunk 4]
```diff
@@ -524,271 +493,36 @@ Configuration ArcGISConfigure
...
                     }
-                    #Check if to add fallbacks in case per node cert is not given or missing.
-
-                    if($Federation -and ($Node.NodeName -ieq $PrimaryServerMachine) -and (($AllNodes | Where-Object { ($_.Role -icontains 'LoadBalancer' -or $_.Role -icontains 'ServerWebAdaptor')}  | Measure-Object).Count -eq 0))
-                    {
-                        if(($AllNodes | Where-Object { $_.Role -icontains 'PortalWebAdaptor'}  | Measure-Object).Count -eq 0)
-                        {
-                            if($PrimaryPortalMachine)
+                    else{
+                        if(@("10.5","10.5.1","10.4.1").Contains($ConfigurationData.ConfigData.Version)){
```
Comment (spitzerr) [inline Modules/ArcGIS/Configuration/ArcGISConfigure.ps1]: Why is it only for this specific versions?

File: Modules/ArcGIS/ArcGIS.psm1 (hunk 1 above)
Comment (shailesh91) [reply Modules/ArcGIS/ArcGIS.psm1]: Yep, Will fix this.

File: Modules/ArcGIS/ArcGIS.psm1 (hunk 2 above)
Comment (shailesh91) [reply Modules/ArcGIS/ArcGIS.psm1]: Yes, you are correct. Will fix this.

File: Modules/ArcGIS/DSCResources/ArcGIS_Portal/ArcGIS_Portal.psm1 (hunk 3 above)
Comment (shailesh91) [reply Modules/ArcGIS/DSCResources/ArcGIS_Portal/ArcGIS_Portal.psm1]: Yes this debug message was split, but I forgot to remove the message. Will remove this.

File: Modules/ArcGIS/Configuration/ArcGISConfigure.ps1 (hunk 4 above)
Comment (shailesh91) [reply Modules/ArcGIS/Configuration/ArcGISConfigure.ps1]: From 10.6 onwards, the server comes with a Self Signed Certificate already installed.

**Comment by shailesh91** [comment]
@spitzerr I will consider making the necessary changes to the federation resource.

**Pull request merged**